server/session.go

```diff
@@ -255,12 +257,24 @@ func (s *Session) isCluster() bool {
 	return s.isProxy() || s.isMultiplex()
 }
 
+func (s *Session) unlockTerminating() {
+	if s == nil {
+		return
+	}
+	s.terminatingLock.Unlock()
+}
+
 // queueOut attempts to send a ServerComMessage to a session write loop; if the send buffer is full,
 // timeout is `sendTimeout`.
 func (s *Session) queueOut(msg *ServerComMessage) bool {
+	defer s.unlockTerminating()
+	if s != nil {
+		s.terminatingLock.Lock()
+	}
 	if s == nil || s.terminating {
```

Comment:
Consider making this the first check in the method.

Reply:
Yup.